### DISCUSSION TITLE
Vitest + React 18: ReferenceError: React is not defined

### DISCUSSION THREAD

**Discussion by jakub-murin** [vitest-dev/vitest#2053] [Q&A] [answered]
I'm attempting to integrate Vitest into my React 18 project for the first time. It seems Vitest cannot handle `.tsx` files not importing `React` explicitely which is supposed to be unnecessary since React 17. The app runs fine but all tests fail with

> ReferenceError: React is not defined

The only way I've found to handle the problem is to `import React from "react";` in every file, which is not acceptable. I'm sure I'm missing something trivial like a config option that would solve it.

I've set up a [minimal project repository on GitHub](https://github.com/zukunftstech/vitest_react18_poc) demonstrating the problem.

**Comment by sheremet-va** [comment, marked as answer]
Use @vitejs/plugin-react

**Reply by jakub-murin** [reply]
I'm already using it:

[package.json](https://github.com/zukunftstech/vitest_react18_poc/blob/main/package.json)
```
  "devDependencies": {

    "@vitejs/plugin-react": "^2.1.0",
  }
```

**Reply by sheremet-va** [reply]
Vitest config overrides Vite config. You don't have plugin there. Please, refer to configuration docs.

**Reply by jakub-murin** [reply]
Awesome, that was it, thank you!

**Reply by kristiandupont** [reply]
I am having issues with this. If I try to use it in my `vitest.config.ts`, I get the error "Type 'PluginOption[]' is not assignable to type 'PluginOption'."

If I try to use it without brackets, like this:
```ts
export default defineConfig({
  plugins: react(),
  test: {
    environment: "jsdom",
  },
});
```

I get a really long and verbose error about incompatible `PluginOption[]` types. What could I be doing wrong?

**Reply by sheremet-va** [reply]
Plugins expects an array.

**Reply by kristiandupont** [reply]
Sure, but if I try that, I get this:
<img width="536" alt="image" src="https://user-images.githubusercontent.com/847742/235975049-16f69f5e-f4a9-4778-80e8-22689c9141f0.png">

**Reply by sheremet-va** [reply]
Looks like you have several vite versions installed. Check your package-lock.

**Reply by kristiandupont** [reply]
That was indeed the problem. I had Vite 4.3.4 but since Vitest depends on 4.1.4, I downgraded to that version.

**Reply by sheremet-va** [reply]
> That was indeed the problem. I had Vite 4.3.4 but since Vitest depends on 4.1.4, I downgraded to that version. 

Vitest doesn't depend on a lower version, this is just how your package manager resolves it.

**Reply by tannerhuynh** [reply]
FYI I have this plugin, but still seeing the same referenceError.

**Reply by pranay-dutta** [reply]
If you are using separate vite.config and vitest config you've to mege them together.
Read it here https://vitest.dev/guide/#configuring-vitest

### Sample file of `vite.config.ts` after merging
```ts
import { defineConfig, mergeConfig } from "vitest/config";
import viteConfig from "./vite.config";

export default mergeConfig(
  viteConfig,
  defineConfig({
    test: {
      environment: "jsdom", //or which ever runtime enviroment you're using
     },
  })
);
```
